### PR TITLE
fix(ivy): prevent ngcc from referencing missing ɵsetClassMetadata

### PR DESCRIPTION
**Pull request by JoostK**
When ngtsc compiles @angular/core, it rewrites core imports to the
r3_symbols.ts file that exposes all internal symbols under their
external name. When creating the FESM bundle, the r3_symbols.ts file
causes the external symbol names to be rewritten to their internal name.

Under ngcc compilations of FESM bundles, the indirection of
r3_symbols.ts is no longer in place such that the external names are
retained in the bundle. Previously, the external name `ɵdefineNgModule`
was explicitly declared internally to resolve this issue, but the
recently added `setClassMetadata` was not declared as such, causing
runtime errors.

Instead of relying on the r3_symbols.ts file to perform the rewrite of
the external modules to their internal variants, the translation is
moved into the `ImportManager` during the compilation itself. This
avoids the need for providing the external name manually.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

ngcc compiled @angular/core would reference `setClassMetadata` by its private name `ɵsetClassMetadata` which wouldn't be available during runtime.

## What is the new behavior?

External references are rewritten to their internal names during compilation, instead of relying on `r3_symbols.ts` as that is not in use when ngcc processes the FESM bundles of @angular/core.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

File: packages/core/src/core_render3_private_export.ts [hunk 1]
```diff
@@ -7,15 +7,12 @@
  */
 
 // clang-format off
-// We need to have `ɵdefineNgModule` defined locally for flat-file ngcc compilation.
-// More details in the commit where this is added.
-import {defineNgModule} from './render3/index';
-export const ɵdefineNgModule = defineNgModule;
```
Comment (petebacondarwin) [inline packages/core/src/core_render3_private_export.ts]: If you remove this then you must update the golden symbols file: https://circleci.com/gh/angular/angular/104307

**Comment by petebacondarwin** [comment]
The source needs formatting: https://circleci.com/gh/angular/angular/104308

**Comment by petebacondarwin** [comment]
It seems that these changes are still not enough to fix the FESM5 situation. The ngcc-CLI integration test is complaining: https://circleci.com/gh/angular/angular/104316.

**Comment by JoostK** [comment]
I figured out why the integration test fails:

The `.d.ts` transform is preferred to run for the FESM2015 bundle, for which no `r3_symbols` file can be located. Hence, the imports that are generated for `@angular/core` are not rewritten at all. Because the symbols are now rewritten to their internal names, they fail to be resolved from `@angular/core` (as expected).

If I force the .d.ts transform to run for ESM2015 bundles, I do end up with a relative import but unfortunately that is of the form:
```
import * as ɵngcc0 from '../esm2015/src/r3_symbols';
```

Makes sense, given that the ESM2015 bundle path was used for searching for r3_symbols. For the `.d.ts` transform we should find the `r3_symbols.d.ts` file instead.

Please mark this PR as WIP for now, I will try to resolve the above issue tomorrow evening.

**Comment by JoostK** [comment]
Alright, I pushed a very dirty implementation to see if it would work, and it does. CircleCI isn't being triggered for some reason but I have seen the integration tests pass.

The seconds commit needs some cleanup as it has currently gone wild. Also, tests are needed to verify that this works as expected.

File: packages/compiler-cli/src/ngtsc/util/src/path.ts [hunk 2]
```diff
@@ -10,10 +10,10 @@
 
 import * as path from 'path';
 
-const TS_DTS_EXTENSION = /(\.d)?\.ts$/;
+const TS_DTS_JS_EXTENSION = /(\.d)?\.ts|\.js$/;
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngtsc/util/src/path.ts]: As it is now, the `$` is only applied to the `\.js` part. You need it in both (e.g. `/(\.d)?\.ts$|\.js$/`).

Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/util/src/path.ts]: Nice catch!

This was actually copied from ngtsc's resolver:

https://github.com/angular/angular/blob/4e4bca6bbcc029a8e7516fa0f01495828f503215/packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts#L20

Back then it was already discussed and we ended up with the correct regex, see https://github.com/angular/angular/pull/25406#discussion_r209052202.

IIRC @alxhub then "optimized" it a bit :) Let me fix both instances when I get home tonight.

Comment (gkalpak) [reply packages/compiler-cli/src/ngtsc/util/src/path.ts]: LoL

If you are too interested in optimizing it (which I don't think matters in this case), you should use non-capturing groups (e.g. `/(?:\.d)?\.ts$|\.js$/`) or better yet use no groups (e.g. `/\.ts$|\.d\.ts$|\.js$/`): https://jsperf.com/gh-ng-27055/1

Comment (alfaproject) [reply packages/compiler-cli/src/ngtsc/util/src/path.ts]: If `.d.ts` ends in `.ts`
Then `/\.ts$|\.js$/` or `/\.[tj]s$/`

Comment (gkalpak) [reply packages/compiler-cli/src/ngtsc/util/src/path.ts]: So very true 🤦‍♂️

Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/util/src/path.ts]: @alfaproject for matching that is true, this is used for replacements though. I'll go with @gkalpak's suggestion.

File: packages/compiler-cli/src/ngtsc/util/src/path.ts [hunk 3]
```diff
@@ -10,10 +10,10 @@
 
 import * as path from 'path';
 
-const TS_DTS_EXTENSION = /(\.d)?\.ts$/;
+const TS_DTS_JS_EXTENSION = /(?:\.d)?\.ts$|\.js$/;
```
Comment (alfaproject) [inline packages/compiler-cli/src/ngtsc/util/src/path.ts]: Should this be exported from a utility class? Duplicating it already proved brittle

File: packages/compiler-cli/src/ngcc/src/packages/bundle.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ts from 'typescript';
+
+/**
+ * A bundle represents the currently compiled entry point format, containing
+ * information that is necessary for compiling @angular/core with ngcc.
+ */
+export interface BundleInfo {
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngcc/src/packages/bundle.ts]: I like the concept of an object to hold this stuff, I am not happy with the name though. `Bundle` is too generic.

Also I still switch back and forth between passing everything into the constructor versus passing everything to the `renderProgram` method versus (our current situation) of passing some things to each.

**Comment by JoostK** [comment]
@mhevery FYI, I'm planning on rebasing this on top of #26906 as it's quite conflicty, and @petebacondarwin's PR should be considered ahead of this one.

**Comment by petebacondarwin** [comment]
And after a review from Alex R, I am currently rewriting some portion of that PR, but it should be ready in a few hours.

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/src/ngcc/src/packages/transformer.ts [hunk 5]
```diff
@@ -70,27 +71,40 @@ export class Transformer {
     const host = ts.createCompilerHost(options);
     const rootDirs = this.getRootDirs(host, options);
     const isCore = entryPoint.name === '@angular/core';
-    const r3SymbolsPath = isCore ? this.findR3SymbolsPath(dirname(entryPointFilePath)) : null;
+    const r3SymbolsPath =
+        isCore ? this.findR3SymbolsPath(dirname(entryPointFilePath), 'r3_symbols.js') : null;
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/src/ngcc/src/packages/transformer.ts]: Yes. For FESMs, the `NgccImportManager` will prevent core imports from being generated if core itself is being compiled.

https://github.com/angular/angular/blob/789d32326b4e95e3a04e77a1f3b2de7c430c1f13/packages/compiler-cli/src/ngcc/src/rendering/ngcc_import_manager.ts#L17-L19

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/src/packages/transformer.ts]: Glad that is sorted :-)

**Comment by petebacondarwin** [comment]
@mhevery - can you review this for `integration` and `core` PullApprove?

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
http://test/OCL:222410910:BASE:222410955:1542817212252:aab09c58

**Pull request closed without merging**